sklearn/utils/estimator_checks.py

```diff
@@ -1339,6 +1344,31 @@ def check_estimators_dtypes(name, estimator_orig):
                 getattr(estimator, method)(X_train)
 
 
+def check_estimators_preserve_dtypes(name, estimator_orig):
+
+    X = np.random.RandomState(0).randn(50, 5)
+    y = np.random.RandomState(0).randn(50)
+
+    for dtype_in, dtype_out in [(np.float32, np.float32),
+                                (np.float64, np.float64),
+                                (np.float16, np.float64)]:
+        X_cast = X.copy().astype(dtype_in)
+        estimator = clone(estimator_orig)
+        set_random_state(estimator)
+        if hasattr(estimator, 'fit_transform'):
+            X_trans = estimator.fit_transform(X_cast, y)
+        elif hasattr(estimator, 'fit') and hasattr(estimator, 'transform'):
+            estimator.fit(X_cast, y)
+            X_trans = estimator.transform(X_cast, y)
+
+        # FIXME: should we check that the dtype of some attributes are the
+        # same than dtype and check that the value of attributes
+        # between 32bit and 64bit are close
+        assert X_trans.dtype == dtype_out, \
+            ('Estimator transform dtype: {} - orginal/expected dtype: {}'
+             .format(X_trans.dtype, dtype_out.__name__))
+
```

Comment:
In the end maybe we should add the check comparing transforms in this PR, say using,
```py
   assert_allclose(X_trans_32, X_trans_64, rtol=1e-2)
```
with a high enough tolerance. Before we start modifying other estimators to pass this.

Reply:
why such a high tolerance ?

Reply:
OK, maybe that was too pessimistic. Maybve  rtol=1e-4 then,  what value would be good you think?

Reply:
I really don't know :/ Ideally it would be around 1e-6, but it's not realistic.
The output precision depends of the algorithm and we probably can't expect machine precision for all algorithms even with the best possible implementation.
I put 1e-5 in the tests of our wrappers of scipy blas but I'm fine with 1e-4